src/Components/Blazor/Build/src/targets/Blazor.MonoRuntime.targets

```diff
@@ -219,49 +219,110 @@
-    </ReadLinesFromFile>
+    <PropertyGroup>
+      <_BlazorReferenceCacheFile>$(BlazorIntermediateOutputPath)blazor.assemblyreferences.cache</_BlazorReferenceCacheFile>
+    </PropertyGroup>
 
     <ItemGroup>
-      <BlazorOutputWithTargetPath Include="@(_BlazorResolvedRuntimeDependencies)">
-        <TargetOutputPath>$(BlazorRuntimeBinOutputPath)%(FileName)%(Extension)</TargetOutputPath>
-      </BlazorOutputWithTargetPath>
+      <_BlazorProjectReference Include="@(ReferencePath->WithMetadataValue('ReferenceSourceTarget', 'ProjectReference'))" />
+      <_BlazorExternalReference Include="@(ReferencePath)" Exclude="@(_BlazorProjectReference)" />
     </ItemGroup>
+
+    <BlazorGetAssemblyReferences Assemblies="@(IntermediateAssembly);@(_BlazorProjectReference)">
+      <Output TaskParameter="AssemblyReferences" ItemName="_ProjectAssemblyReferences" />
+    </BlazorGetAssemblyReferences>
+
+    <Hash ItemsToHash="@(_BlazorProjectReference);@(_ProjectAssemblyReferences);">
+      <Output TaskParameter="HashResult" PropertyName="_BlazorReferenceHash" />
+    </Hash>
+
+    <WriteLinesToFile
+      Lines="$(_BlazorReferenceHash)"
+      File="$(_BlazorReferenceCacheFile)"
+      Overwrite="True"
+      WriteOnlyWhenDifferent="True" />
   </Target>
 
   <Target
-    Name="_ResolveBlazorRuntimeDependencies"
+    Name="_GetBlazorNotLinkedAssemblyClosure"
+    DependsOnTargets="_GenerateBlazorLinkerDescriptor"
     Inputs="$(ProjectAssetsFile);
-            @(IntermediateAssembly);
-            @(_BlazorDependencyInput)"
-    Outputs="$(_BlazorApplicationAssembliesCacheFile)">
+            $(_BlazorReferenceCacheFile);
+            @(_BlazorExternalReference);"
+    Outputs="$(_BlazorLinkerOutputCache)">
+
+   <ItemGroup>
+      <_OldLinkedFile Include="$(BlazorIntermediateLinkerOutputPath)*.dll" />
+      <_OldLinkedFile Include="$(BlazorIntermediateLinkerOutputPath)*.pdb" />
+    </ItemGroup>
+
+    <Delete Files="@(_OldLinkedFile)" />
 
     <!--
-    At this point we have decided not to run the linker and instead to just copy the assemblies
-    from the BCL referenced by the app the nuget package into the _framework/_bin folder.
-    The only thing we need to do here is collect the list of items that will go into _framework/_bin.
+      When running from Desktop MSBuild, DOTNET_HOST_PATH is not set.
+      In this case, explicitly specify the path to the dotnet host.
     -->
-    <ResolveBlazorRuntimeDependencies
-      EntryPoint="@(IntermediateAssembly)"
-      ApplicationDependencies="@(_BlazorDependencyInput)"
-      WebAssemblyBCLAssemblies="@(_WebAssemblyBCLAssembly)">
+    <PropertyGroup Condition=" '$(DOTNET_HOST_PATH)' == '' ">
+      <_DotNetHostDirectory>$(NetCoreRoot)</_DotNetHostDirectory>
+      <_DotNetHostFileName>dotnet</_DotNetHostFileName>
+      <_DotNetHostFileName Condition=" '$(OS)' == 'Windows_NT' ">dotnet.exe</_DotNetHostFileName>
+    </PropertyGroup>
+
+    <BlazorILLink
```

Comment:
This is based on Steve's conversation with the mono team. The linker has a mode which effectively does what our assembly resolution did - it copies things to the output if assemblies are referenced by the root without linking them. This is pretty identical to what we do with our running when not linked on build.

The linker task as it's implemented in our code and in the SDK shells out to a dotnet process. On my machine for the test projects we have checked in, running it takes about 2 sec. This turns out to be the most expensive operation in the build. As @SteveSandersonMS pointed out, users probably turn off the linker for faster inner loop, so running it on every build where it's the most expensive task would defeat the purpose.

This PR attempts to add some build incrementalism to the mix. My reasoning was that if

a) None of the package references have changed (project.assets.json is unchanged)
b) Assembly references for the current project and any of the project references are unchanged

we could say a previously calculated assembly closure is still good. This allows us to run the linker task only if the built assembly has different references. In an ordinary build triggered after you modified your component and none of the references changed, this would no-op this target.

Project references make it a bit more complicated, but I wanted to make sure putting components in a referenced RCL was still performant.

What do you think? Are there any obvious scenarios I missed?